Do domain resolution of 'tcp://' scheme addresses

nicomem opened the pull request:
Before all, thanks for making this library. In my search for one to support systemd sockets, this is the one that I found to have the nicest API.

## Problem

When I try to parse `tcp://localhost:8080` into `Binding`, I get the following error: `BadAddress(AddrParseError(Socket))`.

## Solution

Use `str.to_socket_addrs()` instead of `str.parse()`.

## Impacts & Important notes

`io::Error` is returned by `str.to_socket_addrs()` instead of previously `AddrParseError`, so the `Error` enum value returned had to be modified.

`str.to_socket_addrs()` may return multiple value (e.g. the IPv4 and IPv6 address of `localhost`). Here, I returned the first value of the iterator.

Unrelated to these changes, some tests failed (on `main`), which seems to be because the `nightly` toolchain now does more checks to protect against using an invalid file descriptor.
I have marked those problematic tests as `#[ignore]` to avoid blocking new changes, but should probably be fixed in the future.

nicomem committed:
Resolve domain of 'tcp://' scheme addresses

This allows for example 'tcp://localhost:8080' to be parsed
  into `Binding` instead of returning an error.

Signed-off-by: Nicolas Mémeint <git@nicomem.com>

wiktor-k commented:
Hi @nicomem 👋 

Thanks for your kind words! Ever since I've created this lib all my other network crates use it but it's super nice to hear someone else is also considering using it 😅

Good point with the unit tests: they need adjusting (I can do that in a separate PR).

As for your PR I'm okay with it especially since service binding is modelled after dockerd `-H` argument which also supports domain names.

Do note though that due to the indeterminate nature of DNS resolution order you may not get the socket you're asking for (as in: this may be a footgun).

The [`net.Listen` API](https://pkg.go.dev/net#Listen) (which powers dockerd tcp protocol) has the following disclaimer:

> The address can use a host name, but this is not recommended, because it will create a listener for at most one of the host's IP addresses.

So I think it'd be good to at lrsst add something like this to the docs here too.

Thanks for your time!

nicomem commented:
I saw that `TcpListener` already takes a `ToSocketAddrs`, so we do not have to select the first and instead forward all the `SockAddr` we get to it.
This is better than previously because `TcpListener::bind` tests all `SockAddr` until one succeeds the binding.

As per the disclaimer you mentioned, using a hostname still may be a footgun, but at least we are not worse than what `std` does.

To do this, I could not use `Binding` as-is, since it could only hold one `SockAddr`.
I added another enum item `Sockets(Vec<SockAddr>)`, which should be less breaking than changing `Socket(SockAddr)`.

I saw that the test `parse_tcp_fail` failed only on the Windows CI. I presume that Windows authorizes a socket address in the format `:8080`.
So I changed the address to `tcp://::8080` (it should be rejected everywhere), and added another fail case with a hostname that should be unknown.

wiktor-k commented:
I agree with everything you've said expect this tiny bit:

> I added another enum item Sockets(Vec<SockAddr>), which should be less breaking than changing Socket(SockAddr).

Since it's breaking anyway I think having just the plural will make the code simpler (no need to check for len etc).

Most of the clients go from Binding to a Listener directly ([like so](https://github.com/wiktor-k/ssh-agent-lib/pull/56/files#diff-4e7fb2261df3c70042fe7740d724183d99f9fce4ec1787b84c0073d08689f55dR204)) so it shouldn't affect them.

(I'll bump the major version anyway).

What do you think about it?

The code is good as it is anyway so I'm okay with merging it and then fixing it myself along with the tests.

wiktor-k commented:
Okay. Check out https://github.com/wiktor-k/service-binding/pull/36 it should resolve all failing tests. If you're ok with #36 we can merge that and then rebase this on top so that the `#[ignore]`s disappear... phew :sweat_smile:

nicomem commented:
Good job fixing the tests 👍

> Since it's breaking anyway I think having just the plural will make the code simpler (no need to check for len etc).

I agree, this will be simpler.
I'll do these changes and rebase the branch after your merge.

wiktor-k commented:
Okay. Looks great 👍 

I've merged it now and after a couple of small tweaks I'll release a new version this week.

Thanks a lot for the PR and have a nice day! 👋

nicomem commented:
Thank you, have a nice day too! :wave:

wiktor-k commented:
Okay, it's done: https://github.com/wiktor-k/service-binding/releases/tag/v3.0.0 and [released on crates.io](https://crates.io/crates/service-binding).

Thanks again for implementing it! :bow: See you later!
